src/heretic/model.py

```
@@ -103,27 +122,86 @@ def get_layers(self) -> ModuleList:
+        if getattr(gen_config, "top_p", None) not in (None, 1.0):
+            sampling_fields.append("top_p")
+        if getattr(gen_config, "typical_p", None) not in (None, 1.0):
+            sampling_fields.append("typical_p")
+
+        if sampling_fields and not bool(getattr(gen_config, "do_sample", False)):
+            gen_config.do_sample = True
+
+        with suppress(Exception):
+            gen_config.validate()
+
+    def get_layer_type(self, layer_index: int) -> str:
+        """
+        Detect the type of a layer.
+        Returns: 'attention', 'mamba', 'conv', 'hybrid', or 'unknown'
+        """
+        layer = self.get_layers()[layer_index]
+
+        layer_types = []
+
+        # Check for attention
+        if hasattr(layer, "self_attn"):
+            layer_types.append("attention")
+
+        # Check for Mamba/SSM
+        if hasattr(layer, "mamba"):
+            layer_types.append("mamba")
+
+        # Check for Conv
+        if hasattr(layer, "conv"):
+            layer_types.append("conv")
+
+        if len(layer_types) == 0:
+            return "unknown"
+        elif len(layer_types) == 1:
+            return layer_types[0]
+        else:
+            # Layer has multiple operator types
+            return "hybrid"
+
     def get_layer_matrices(self, layer_index: int) -> dict[str, list[Tensor]]:
         layer = self.get_layers()[layer_index]
 
-        matrices = {}
+        matrices: dict[str, list[Tensor]] = {}
 
         def try_add(component: str, matrix: Any):
+            if matrix is None:
+                return
+
             # Handle Triton tensors (e.g., from MXFP4 quantization) by extracting
             # the underlying PyTorch tensor via the .data attribute.
             if hasattr(matrix, "data") and torch.is_tensor(matrix.data):
                 matrix = matrix.data
 
-            assert torch.is_tensor(matrix)
+            if not torch.is_tensor(matrix):
+                return
```

Why did you change this? We suppress exceptions everywhere now, right?